plan/match_property.go

```diff
@@ -16,23 +16,57 @@ package plan
 import (
 	"math"
 
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/model"
 	"github.com/pingcap/tidb/util/types"
 )
 
+func tryToAddUnionScan(txn kv.Transaction, conds []expression.Expression, resultPlan PhysicalPlan) PhysicalPlan {
+	if txn != nil && !txn.IsReadOnly() {
+		us := &PhysicalUnionScan{
+			Condition: expression.ComposeCNFCondition(conds),
+		}
+		us.SetChildren(resultPlan)
+		us.SetSchema(resultPlan.GetSchema())
+		return us
+	}
+	return resultPlan
+}
+
 // matchProperty implements PhysicalPlan matchProperty interface.
 func (ts *PhysicalTableScan) matchProperty(prop *requiredProperty, infos ...*physicalPlanInfo) *physicalPlanInfo {
 	rowCount := float64(infos[0].count)
 	cost := rowCount * netWorkFactor
+	if prop.limit != nil {
+		cost = float64(prop.limit.Count+prop.limit.Offset) * netWorkFactor
+	}
 	if len(prop.props) == 0 {
-		return enforceProperty(prop, &physicalPlanInfo{p: ts, cost: cost, count: infos[0].count})
+		p := tryToAddUnionScan(ts.txn, ts.conditions, ts)
```

Comment:
"try to add union scan" will check if the txn is read-only. If it is, it will add union scan for it.